guides/developers-guide/core/developing-graphql.md

```diff
@@ -0,0 +1,187 @@
+
+> By default, InMemoryCache will attempt to use the commonly found primary keys of `id` and `_id` for the unique identifier if they exist
+
+This does not specifically require global uniqueness since it also uses `__typename`, but because Relay does, we've opted to ensure IDs are globally unique.
+
+In most cases, actual internal data IDs are in MongoDB collections, so they are guaranteed unique within the collection, but not among all collections. To add that extra layer of uniqueness, we concatenate the namespace with the internal ID, and then to keep it looking like a "not human‐readable" ID, we base64 encode.
+
+To convert internal IDs to opaque UUIDs, we first prefix them with "reaction/\<namespace\>" and then base64 encode them. The primary transformation functions that handle this are in the `api-utils` package.
+
+The GraphQL resolver functions are the place where ID encoding and decoding happens. They then call out to plugin functions that deal exclusively with internal IDs. Any IDs returned by such functions must also be transformed before returning them, although this typically and preferably happens in a type resolver.
+
+## Using MongoDB in GraphQL
+
+The source-of-truth database for most data used by GraphQL resolvers is currently MongoDB. The MongoDB Node driver collections API is documented here: http://mongodb.github.io/node-mongodb-native/3.1/api/Collection.html
+
+All collections registered by any plugin are available on the `context` object as `context.collections`.
+
+### Checking whether an operation was successful
+
+For `insertOne` or `insertMany`:
+
+```js
+const { insertedCount } = await SomeCollection.insertOne(/* ... */);
+if (insertedCount === 0) {
+  // throw Error or otherwise handle failure
+}
+```
+
+For `updateOne` or `updateMany`:
+
+```js
+const { modifiedCount } = await SomeCollection.updateOne(/* ... */);
+if (modifiedCount === 0) {
+  // throw Error or otherwise handle failure
+}
+```
+
+For `deleteOne`:
+
+```js
+const { deletedCount } = await SomeCollection.updateOne(/* ... */);
+if (deletedCount === 0) {
+  // throw Error or otherwise handle failure
+}
+```
+
+Keep in mind that sometimes a zero `modifiedCount` or `deletedCount` might be because nothing matched your query, and depending on the situation, this may not be an error. If you foresee this situation, you can opt to check `matchedCount` instead.
+
+## Optimizing GraphQL resolvers
+
+Because of the way GraphQL queries and relationships work, sometimes a query will include something like this:
+
+```graphql
+{
+  order {
+    shop {
+      _id
+    }
+  }
+}
+```
+
+Normally the `shop` relationship would result in a database query, but if `order` already has a `shopId` property, we can actually skip the database lookup because the client has requested only the `_id` property. There is a utility function that helps with this: `optimizeIdOnly`. Check out the `Query.viewer` resolver for an example of how to use it.
```

Comment:
Maybe we should add link to the `Query.viewer` resolver here.

Reply:
Where is that located?

Reply:
@Akarshit second request for clarification here

Reply:
Here is the link: https://github.com/reactioncommerce/api-plugin-accounts/blob/9306d27721a8fe5edf1793d8def4999d9395ddbc/src/resolvers/Query/viewer.js#L17